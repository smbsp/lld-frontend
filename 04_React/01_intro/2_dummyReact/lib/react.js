function react(component){
    // calling that function.
    let dom = component();
    console.log(".......optimizing the changes.");
    return dom;
}