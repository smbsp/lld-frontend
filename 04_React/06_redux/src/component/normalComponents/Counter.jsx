import React from 'react'
import { useState } from 'react'

function Counter() {
    const [count, setCount] = useState(0);

    // Bussiness Logic
    const handleIncrement = () => {
        setCount(count + 1);
    }

    const handleDecrement = () => {
        setCount(count - 1);
    }

    return (
        <>
            <button onClick={handleIncrement}>+</button>
            <h3>{count}</h3>
            <button onClick={handleDecrement}>-</button>
        </>
    )
}

export default Counter