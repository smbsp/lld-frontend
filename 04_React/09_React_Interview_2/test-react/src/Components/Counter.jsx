import React from 'react'
import { useState } from 'react'

function Counter() {
    let [count, changeCount] = useState(0);
    let increment = () => {
        changeCount(count + 1);
    }

    let decrement = () => {
        changeCount(count - 1);
    }
    return (
        <div className='container'>
            <button onClick={decrement}>-</button>
            <p>Count: {count} </p>
            <button onClick={increment}>+</button>
        </div>
    )
}

export default Counter