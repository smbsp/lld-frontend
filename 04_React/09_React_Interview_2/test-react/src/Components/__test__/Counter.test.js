// VDOM -> react-tesing-library
import { render, screen, fireEvent } from '@testing-library/react';
import Counter from '../Counter';

/**
 * Unit testing : testing a component in isolation
  * by testing that in all the possible scinearios
  * We are using react-testing-lib to emulate rendering of that component
 * */

describe("Counter component test cases", () => {

    test("snapshot for counter", () => {
        const {asFragment} = render(<Counter />);
        expect(asFragment()).toMatchSnapshot();
    })

    test("intial state check", () => {
        render(<Counter />);

        const countText = screen.getByText("Count: 0");
        const plusText = screen.getByText("+");
        const minusText = screen.getByText("-");

        //assertions.
        expect(countText).toBeInTheDocument();
        expect(plusText).toBeInTheDocument();
        expect(minusText).toBeInTheDocument();
    })

    test("increment state check", () => {
        render(<Counter />);
        const plusText = screen.getByText("+");

        // Emulate one + click.
        fireEvent.click(plusText);

        //Assertion.
        expect(screen.getByText("Count: 1")).toBeInTheDocument();

        // Emulate second + click.
        fireEvent.click(plusText);

        //Assertion.
        expect(screen.getByText("Count: 2")).toBeInTheDocument();

        // Emulate third + click.
        fireEvent.click(plusText);

        //Assertion.
        expect(screen.getByText("Count: 3")).toBeInTheDocument();
    })

    test("decrement state check", () => {
        render(<Counter />);
        const minusText = screen.getByText("-");

        // Emulate one - click.
        fireEvent.click(minusText);

        //Assertion.
        expect(screen.getByText("Count: -1")).toBeInTheDocument();

        // Emulate second - click.
        fireEvent.click(minusText);

        //Assertion.
        expect(screen.getByText("Count: -2")).toBeInTheDocument();

        // Emulate third - click.
        fireEvent.click(minusText);

        //Assertion.
        expect(screen.getByText("Count: -3")).toBeInTheDocument();
    })
})