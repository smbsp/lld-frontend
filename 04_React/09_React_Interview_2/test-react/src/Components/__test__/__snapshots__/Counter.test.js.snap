// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component test cases snapshot for counter 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <button>
      -
    </button>
    <p>
      Count: 0 
    </p>
    <button>
      +
    </button>
  </div>
</DocumentFragment>
`;
